go-from-rust.md

```diff
@@ -0,0 +1,89 @@
+
+A few factors contributing to learnability are:
+
+* Go is small. Lots of languages try to be small, Go really is small; really small. (This is mostly a good thing and I'm impressed with the discipline this must have taken).
+* The standard library is good (and again, small). Finding and using libraries from the ecosystem is very easy.
+* There is very little in the language which is not in other languages. Go takes lots of bits from other established languages, polishes them, and puts them together nicely. It tries very hard to avoid novelty.
+
+## Boilerplate
+
+Go code becomes very repetitive very quickly. It is missing any mechanism like macros or generics for reducing repetition (interfaces are nice for abstraction, but don't work so well to reduce code duplication). I often end up with lots of functions, identical except for types.
+
+Error handling also causes repetition. Many functions have more `if err != nil { return err }` boilerplate than interesting code.
+
+Using generics or macros to reduce boilerplate is sometimes criticised for making code easy to write at the expense of making it harder to read. But I found the opposite with Go. It is quick and easy to copy and paste code, but reading Go code is frustrating because you have to ignore so much of it or search for subtle differences.
+
+## Things I like
+
+* Compile times. Definitely quick; definitely a lot quicker than Rust. But not actually as quick as I was expecting (it feels to me to be similar to C/C++ for medium to large projects, I was kind of expecting instant).
+* Go routines and channels. Having lightweight syntax for spawning Go routines and using channels is really nice. It really shows the power of syntax that such a small detail makes concurrent programming feel so much nicer than in other languages.
+* Interfaces. They're not very sophisticated but they are easy to understand and use, and useful in a lot of places.
+* `if ...; ... { }` syntax. Being able to restrict the scope of variables to the body of `if` statements is nice. This has a similar effect to `if let` in Swift and Rust, but is more general purpose (Go does not have pattern matching like Swift and Rust, so it cannot use `if let`.
+* Testing and doc comments are easy to use.
+* The `Go` tool is nice - having everything in one place rather than requiring multiple tools on the command line.
+* Having a garbage collector! Not thinking about memory management really does make programming easier.
+* Varargs
+
+## Things I did not like
+
+In no particular order.
+
+* `nil` slices - `nil`,  a `nil` slice, and an empty slice are all different. I'm pretty sure you only need two of those things, not three.
+* No first class enums. Using constants feels backwards.
+* Disallowing import cycles. This really limits how useful packages are for modularising a project, since it encourages putting lots of files in a package.
```

Comment:
That's the approach usually taken by go users, but I actually solve it by using lots of really small packages. Whenever there is an import cycle that just means there should be another package.